File: clients/rust/src/instructions/manage_integration/manage_integration.rs
use crate::{
    derive_controller_authority_pda, derive_permission_pda,
    generated::{instructions::ManageIntegrationBuilder, types::IntegrationStatus},
};
use solana_instruction::Instruction;
use solana_pubkey::Pubkey;

/// Instruction generation for managing a integration account

pub fn create_manage_integration_instruction(
    controller: &Pubkey,
    authority: &Pubkey,
    integration: &Pubkey,
    status: IntegrationStatus,
    rate_limit_slope: u64,
    rate_limit_max_outflow: u64,
) -> Instruction {
    let calling_permission_pda = derive_permission_pda(controller, &authority);
    let controller_authority = derive_controller_authority_pda(controller);

    ManageIntegrationBuilder::new()
        .status(status)
        .rate_limit_slope(rate_limit_slope)
        .rate_limit_max_outflow(rate_limit_max_outflow)
        .controller(*controller)
        .controller_authority(controller_authority)
        .authority(*authority)
        .permission(calling_permission_pda)
        .integration(*integration)
        .program_id(crate::SVM_ALM_CONTROLLER_ID)
        .instruction()
}

File: clients/ts/src/instructions/manageIntegration.ts
import { Address, TransactionSigner, OptionOrNullable } from "@solana/kit";
import { getManageIntegrationInstruction } from "../generated/instructions/manageIntegration";
import {
  deriveControllerAuthorityPda,
  derivePermissionPda,
} from "../pdas";
import { SVM_ALM_CONTROLLER_PROGRAM_ADDRESS } from "../generated";
import type { IntegrationStatusArgs } from "../generated/types";

/**
 * Instruction generation for managing an integration account
 */
export async function createManageIntegrationInstruction(
  controller: Address,
  authority: TransactionSigner,
  integration: Address,
  status: OptionOrNullable<IntegrationStatusArgs>,
  rateLimitSlope: OptionOrNullable<number | bigint>,
  rateLimitMaxOutflow: OptionOrNullable<number | bigint>,
  description?: OptionOrNullable<Uint8Array>
) {
  const permissionPda = await derivePermissionPda(controller, authority.address);
  const controllerAuthority = await deriveControllerAuthorityPda(controller);

  return getManageIntegrationInstruction({
    controller,
    controllerAuthority,
    authority,
    permission: permissionPda,
    integration,
    programId: SVM_ALM_CONTROLLER_PROGRAM_ADDRESS,
    status,
    description: description ?? null,
    rateLimitSlope,
    rateLimitMaxOutflow,
  });
}

